lib_neutral_prompt/cfg_denoiser_hijack.py

```diff
@@ -212,27 +212,121 @@ def filter_abs_top_k(vector: torch.Tensor, k_ratio: float) -> torch.Tensor:
 
-        return sampler
+    forge_sampler_hijacker = hijacker.ModuleHijacker.install_or_get(
+        module=forge_sampler,
+        hijacker_attribute='__forge_sample_hijacker',
+        on_uninstall=script_callbacks.on_script_unloaded,
+    )
+
+    @forge_sampler_hijacker.hijack('forge_sample')
+    def forge_sample(self, denoiser_params, cond_scale, cond_composition, original_function):
+        if not global_state.is_enabled:
+            return original_function(self, denoiser_params, cond_scale, cond_composition)
+
+        model = self.inner_model.inner_model.forge_objects.unet.model
+        control = self.inner_model.inner_model.forge_objects.unet.controlnet_linked_list
+        extra_concat_condition = self.inner_model.inner_model.forge_objects.unet.extra_concat_condition
+        x = denoiser_params.x
+        timestep = denoiser_params.sigma
+        model_options = self.inner_model.inner_model.forge_objects.unet.model_options
+        seed = self.p.seeds[0]
+
+        uncond = forge_sampler.cond_from_a1111_to_patched_ldm(denoiser_params.text_uncond)
+        conds = forge_sampler.cond_from_a1111_to_patched_ldm_weighted(denoiser_params.text_cond, cond_composition)
+        conds += uncond
+
+        denoised = []
+
+        for current_cond in conds:
+            cond = [current_cond]
+            cond[0]['strength'] = 1.0
+
+            if extra_concat_condition is not None:
+                image_cond_in = extra_concat_condition
+            else:
+                image_cond_in = denoiser_params.image_cond
+
+            if isinstance(image_cond_in, torch.Tensor):
+                if image_cond_in.shape[0] == x.shape[0] \
+                        and image_cond_in.shape[2] == x.shape[2] \
+                        and image_cond_in.shape[3] == x.shape[3]:
+                    cond[0]['model_conds']['c_concat'] = CONDRegular(image_cond_in)
+
+            if control is not None:
+                cond[0]['control'] = control
+
+            for modifier in model_options.get('conditioning_modifiers', []):
+                model, x, timestep, _, cond, cond_scale, model_options, seed = modifier(model, x, timestep, None, cond, cond_scale, model_options, seed)
+
+            model_options["disable_cfg1_optimization"] = True
+
+            result = sampling_function(model, x, timestep, None, cond, 1.0, model_options, seed)
+            denoised.append(result)
+
+        cond_indices = cond_composition[0]
+        prompt = global_state.prompt_exprs[0]
 
-    sampler.model_wrap_cfg.combine_denoised = functools.partial(
-        combine_denoised_hijack,
-        original_function=sampler.model_wrap_cfg.combine_denoised
+        # B, C, H, W
+        denoised_uncond = denoised[-1]
+
+        # N, B, C, H, W
+        denoised_conds = torch.stack(denoised[:-1], dim=0)
+
+        # N, 1, 1, 1, 1
+        weights = torch.tensor([ weight for (_, weight) in cond_indices ], device=denoised_uncond.device)
+        weights /= weights.abs().sum()
```

Comment:
prompt weighting works properly now in the basic case

I also used L1 norm instead of arithmetic mean for the weighted average -- it's equivalent when the weights are all positive, but also properly normalizes (and avoids potential division by zero) when some weights are negative

I think it's still incorrect in the case of bracketed/nested prompts... will test more, but I think doing a separate weighted average for each bracketed group is what's needed

Reply:
I think weight normalization via tree traversal is a good feature for this extension in the future, but a distraction from the purpose of this PR, so I dropped it for now

Reply:
No problem, although I want to make it clear I think L1 could be very useful to maintain a sensible latent scale. Without L1, using multiple AND can be thought of as multiplying the CFG by the number of prompts, which is not really user friendly.